federation/planner.go

```diff
@@ -0,0 +1,318 @@
+		if selection.Name != "__typename" {
+			return nil, fmt.Errorf("unexpected selection %s on union", selection.Name)
+		}
+		plan.SelectionSet.Selections = append(plan.SelectionSet.Selections, selection)
+	}
+
+	// We expect at most one suplan per type since the query is flattened
+	seenFragments := make(map[string]struct{})
+
+	for _, fragment := range selectionSet.Fragments {
+		// Enforce flattened schema.
+		if _, ok := seenFragments[fragment.On]; ok {
+			return nil, fmt.Errorf("reused fragment %s, expected flattened query", fragment.On)
+		}
+		seenFragments[fragment.On] = struct{}{}
+
+		// All fragments must be on concrete types
+		typ, ok := typ.Types[fragment.On]
+		if !ok {
+			return nil, fmt.Errorf("unexpected fragment on %s for typ %s", fragment.On, typ.Name)
+		}
+
+		// Create a plan for all fragment types
+		concretePlan, err := e.plan(typ, fragment.SelectionSet, service)
+		if err != nil {
+			return nil, err
+		}
+
+		// Query the fields known to the current with a local fragment.
+		plan.SelectionSet.Fragments = append(plan.SelectionSet.Fragments, &graphql.Fragment{
+			On:           typ.Name,
+			SelectionSet: concretePlan.SelectionSet,
+		})
+
+		// Make subplans conditional on the current type.
+		for _, subPlan := range concretePlan.After {
+			subPlan.Path = append(subPlan.Path, PathStep{Kind: KindType, Name: typ.Name})
+			plan.After = append(plan.After, subPlan)
+		}
+	}
+
+	return plan, nil
+}
+
+func (e *Planner) plan(typIface graphql.Type, selectionSet *graphql.SelectionSet, service string) (*Plan, error) {
+	switch typ := typIface.(type) {
+	case *graphql.NonNull:
+		return e.plan(typ.Type, selectionSet, service)
+
+	case *graphql.List:
+		return e.plan(typ.Type, selectionSet, service)
+
+	case *graphql.Object:
+		return e.planObject(typ, selectionSet, service)
+
+	case *graphql.Union:
+		return e.planUnion(typ, selectionSet, service)
+
+	default:
+		return nil, fmt.Errorf("bad typ %v", typIface)
+	}
+}
+
+// reversePaths reverses all paths in the plan and its subplans.
+//
+// Building reverse plans is easier with append, this cleans up the mess.
+func reversePaths(p *Plan) {
+	for i := 0; i < len(p.Path)/2; i++ {
+		j := len(p.Path) - 1 - i
+		p.Path[i], p.Path[j] = p.Path[j], p.Path[i]
+	}
+	for _, p := range p.After {
+		reversePaths(p)
+	}
+}
+
+func (e *Planner) planRoot(query *graphql.Query) (*Plan, error) {
+	var schema graphql.Type
+	switch query.Kind {
+	case "query":
+		schema = e.schema.Schema.Query
+	case "mutation":
+		schema = e.schema.Schema.Mutation
+	default:
+		return nil, fmt.Errorf("unknown query kind %s", query.Kind)
+	}
+
+	flattened, err := e.flattener.flatten(query.SelectionSet, schema)
+	if err != nil {
+		return nil, err
+	}
+
+	p, err := e.plan(schema, flattened, "gateway-coordinator-service")
+	if err != nil {
+		return nil, err
+	}
+
+	if query.Kind == "mutation" {
+		if len(p.After) > 1 {
+			return nil, errors.New("only support 1 mutation step to maintain ordering")
```

Comment:
I'm not sure what "maintain ordering" means in this context.  Maybe add a comment explaining what problem this avoids?

Reply:
Added a comment, but it makes sure that mutations run on separate graphql servers won't be run out of order